File: constructor/Constructor.java
// Parent class with constructor
class Piece {
    String name;

    // Default constructor
    Piece() {
        this.name = "Generic Piece";
    }

    // Parameterized constructor
    Piece(String name) {
        this.name = name;
    }

    void display() {
        System.out.println("Piece: " + name);
    }
}

// Child class with constructor and inheritance
class Knight extends Piece {
    int moves;

    // Constructor with super call
    Knight(String name, int moves) {
        super(name); // Calls parent constructor
        this.moves = moves;
    }

    void display() {
        System.out.println("Knight: " + name + ", Moves: " + moves);
    }
}

public class ChessGame {
    public static void main(String[] args) {
        Piece genericPiece = new Piece(); // Default constructor
        genericPiece.display(); // Outputs: Piece: Generic Piece

        Knight knight = new Knight("White Knight", 2); // Parameterized constructor
        knight.display(); // Outputs: Knight: White Knight, Moves: 2
    }
}

File: constructor/constructor.go
package main

import "fmt"

// Struct representing a base piece
type Piece struct {
    Name string
}

// Factory function as a constructor
func NewPiece(name string) *Piece {
    return &Piece{Name: name}
}

// Struct representing a knight
type Knight struct {
    Piece
    Moves int
}

// Factory function as a constructor for Knight
func NewKnight(name string, moves int) *Knight {
    return &Knight{Piece: Piece{Name: name}, Moves: moves}
}

func (p *Piece) Display() string {
    return "Piece: " + p.Name
}

func (k *Knight) Display() string {
    return "Knight: " + k.Name + ", Moves: " + fmt.Sprint(k.Moves)
}

func main() {
    genericPiece := NewPiece("Generic Piece")
    fmt.Println(genericPiece.Display()) // Outputs: Piece: Generic Piece

    knight := NewKnight("White Knight", 2)
    fmt.Println(knight.Display()) // Outputs: Knight: White Knight, Moves: 2
}